Source/Core/Core/HW/BBA-TAP/TAP_Win32.cpp

```diff
@@ -241,56 +259,162 @@ bool CEXIETHERNET::IsActivated()
+		if (completion_key == reinterpret_cast<ULONG_PTR>(self))
+		{
+			// Calculate the address of the containing structure, ASyncIORequest, from the address of the overlapped member.
+			AsyncIORequest* io_hdr = CONTAINING_RECORD(overlapped, AsyncIORequest, overlapped);
+			if (io_hdr->type == AsyncIORequest::IO_PACKET_TYPE_READ)
+			{
+				// Pass the size back to the hardware interface, and fire the interrupt.
+				_dbg_assert_(SP1, read_io_active);
+				if (self->readEnabled.load())
+				{
+					DEBUG_LOG(SP1, "Received %u bytes\n: %s", bytes_transferred, ArrayToString(self->mRecvBuffer, bytes_transferred, 0x10).c_str());
+					self->mRecvBufferLength = bytes_transferred;
+					self->RecvHandlePacket();
+				}
 
-	self->RecvHandlePacket();
+				// Re-use the same read buffer, since it's been copied into HW memory, and queue another read request.
+				read_io_active = QueueRead(self, &read_io_hdr);
+			}
+			else if (io_hdr->type == AsyncIORequest::IO_PACKET_TYPE_WRITE)
+			{
+				// Write buffers are heap-allocated, so must be freed afterwards.
+				// See the note in CEXIETHERNET::SendFrame for why.
+				// The cast here is necessary because it was allocated as a u8[].
+				delete[] reinterpret_cast<u8*>(io_hdr);
+			}
+		}
+		else
+		{
+			// Cancel any outstanding requests from both this thread (reads), and the CPU thread (writes).
+			CancelIoEx(self->mHAdapter, nullptr);
+			while (true)
+			{
+				// Keep dequeuing packets from the queue until it is empty.
+				// This will include the read request queued above, or any outstanding writes.
+				result = GetQueuedCompletionStatus(self->mHCompletionPort, &bytes_transferred, &completion_key, &overlapped, 0);
+				if (!result && !overlapped)
+					break;
+
+				AsyncIORequest* io_hdr = CONTAINING_RECORD(overlapped, AsyncIORequest, overlapped);
+				if (io_hdr->type == AsyncIORequest::IO_PACKET_TYPE_WRITE)
+				{
+					// Free heap buffers for write requests.
+					// The cast here is necessary because it was allocated as a u8[].
+					delete[] reinterpret_cast<u8*>(io_hdr);
+				}
+			}
+
+			return;
+		}
+	}
 }
 
-bool CEXIETHERNET::RecvInit()
+bool CEXIETHERNET::SendFrame(u8* frame, u32 size)
 {
-	// Set up recv event
+	DEBUG_LOG(SP1, "SendFrame %u bytes:\n%s",
+		size, ArrayToString(frame, size, 0x10).c_str());
+
+	// We can't send a frame without any payload.
+	if (!size)
+		return false;
 
-	if ((mHRecvEvent = CreateEvent(nullptr, false, false, nullptr)) == nullptr)
+	// Allocate an IO request, header + data immediately afterwards. This is a little messy, but since our
+	// payload size can vary, the only alternative is two allocations. So in memory, we have [hdr][payload].
+	// The reason for this ugliness is because we have to keep both the buffer and OVERLAPPED structure around
+	// until the kernel has finished with it, and delivers us a completion packet. The memory is then freed
+	// afterwards, in ReadThreadHandler.
+	u8* io_buffer = new u8[sizeof(AsyncIORequest) + size];
```

Comment:
Use vector or something else to avoid raw allocations. Or better yet, refactor the request to handle the correctl sizing data internally. There's absolutely no need for raw allocations and casts like this